File: ai-service/app/core/logger.py
"""
Enterprise CX Guardian AI — Centralized Structured Logger
==========================================================

7 Logging Categories:
  1. Authentication  — login, logout, token events, auth failures
  2. AI Requests     — Groq API calls, model, tokens, latency
  3. Database        — connections, CRUD, health, fallback
  4. RAG Retrieval   — document fetch, similarity scores, chunk counts
  5. Tool Calls      — sentiment, reasoning, recommendation, analyze
  6. Errors          — domain errors, validation, unexpected failures
  7. Response Time   — HTTP requests, per-route latency, p-level alerts

All log lines share a uniform format:
  [YYYY-MM-DD HH:MM:SS] [LEVEL   ] [ai_microservice] [CATEGORY] message
"""

import logging
import sys
import time
from datetime import datetime, timezone
from functools import wraps
from typing import Optional, Any, Callable
from app.core.config import settings

# ══════════════════════════════════════════════════════════════════
# Log Format
# ══════════════════════════════════════════════════════════════════

LOG_FORMAT = (
    "[%(asctime)s] "
    "[%(levelname)-8s] "
    "[%(name)s] "
    "%(message)s"
)
DATE_FORMAT = "%Y-%m-%d %H:%M:%S"

# Latency thresholds for automated WARN elevation (ms)
LATENCY_WARN_MS  = 1000     # 1 second
LATENCY_ERROR_MS = 5000     # 5 seconds


def setup_logger(name: str = "ai_microservice") -> logging.Logger:
    """
    Creates and returns a named logger backed by a stdout StreamHandler.
    Log level driven by settings.LOG_LEVEL (default: INFO).
    Idempotent — safe to call multiple times.
    """
    log = logging.getLogger(name)
    log.setLevel(getattr(logging, settings.LOG_LEVEL.upper(), logging.INFO))

    if not log.handlers:
        handler   = logging.StreamHandler(sys.stdout)
        formatter = logging.Formatter(LOG_FORMAT, datefmt=DATE_FORMAT)
        handler.setFormatter(formatter)
        log.addHandler(handler)

    return log


# Shared application-wide logger instance
logger = setup_logger()


# ══════════════════════════════════════════════════════════════════
# 1. Authentication Logging
# ══════════════════════════════════════════════════════════════════

def log_auth_attempt(email: str, method: str = "password") -> None:
    """Logs a login attempt (before outcome is known)."""
    logger.info(
        f"[Auth] LOGIN ATTEMPT | email: {_mask_email(email)} | method: {method}"
    )


def log_auth_success(user_id: str, email: str, token_type: str = "access") -> None:
    """Logs a successful authentication event."""
    logger.info(
        f"[Auth] ✓ LOGIN SUCCESS | user_id: {user_id} "
        f"| email: {_mask_email(email)} | token: {token_type}"
    )


def log_auth_failure(email: str, reason: str, ip: str = "unknown") -> None:
    """Logs a failed authentication event."""
    logger.warning(
        f"[Auth] ✗ LOGIN FAILED | email: {_mask_email(email)} "
        f"| reason: {reason} | ip: {ip}"
    )


def log_auth_logout(user_id: str) -> None:
    """Logs a logout/token revocation event."""
    logger.info(f"[Auth] LOGOUT | user_id: {user_id}")


def log_token_refresh(user_id: str, success: bool, reason: str = "") -> None:
    """Logs a JWT refresh token usage event."""
    status = "✓ SUCCESS" if success else f"✗ FAILED ({reason})"
    logger.info(f"[Auth] TOKEN REFRESH | user_id: {user_id} | {status}")


def log_token_invalid(reason: str, ip: str = "unknown") -> None:
    """Logs an invalid/expired/forged JWT token event."""
    logger.warning(
        f"[Auth] INVALID TOKEN | reason: {reason} | ip: {ip}"
    )


def log_register(email: str, user_id: str) -> None:
    """Logs a new user registration event."""
    logger.info(
        f"[Auth] REGISTER | user_id: {user_id} | email: {_mask_email(email)}"
    )


# ══════════════════════════════════════════════════════════════════
# 2. AI Request Logging
# ══════════════════════════════════════════════════════════════════

def log_ai_request_start(
    conversation_id: str,
    model: str,
    history_turns: int,
    prompt_chars: int
) -> None:
    """Logs the start of a Groq AI request."""
    logger.info(
        f"[AI Request] → GROQ | conv_id: {conversation_id} "
        f"| model: {model} | history: {history_turns} turns "
        f"| prompt: {prompt_chars} chars"
    )


def log_ai_request_success(
    conversation_id: str,
    model: str,
    prompt_tokens: int,
    completion_tokens: int,
    total_tokens: int,
    latency_ms: float
) -> None:
    """Logs a successful Groq AI response with full telemetry."""
    throughput = (total_tokens / latency_ms * 1000) if latency_ms > 0 else 0
    level = (
        logging.ERROR   if latency_ms >= LATENCY_ERROR_MS else
        logging.WARNING if latency_ms >= LATENCY_WARN_MS  else
        logging.INFO
    )
    logger.log(
        level,
        f"[AI Request] ← GROQ OK | conv_id: {conversation_id} "
        f"| model: {model} "
        f"| tokens: prompt={prompt_tokens} completion={completion_tokens} total={total_tokens} "
        f"| latency: {latency_ms:.0f}ms "
        f"| throughput: {throughput:.0f} tok/s"
    )


def log_ai_request_error(
    conversation_id: str,
    model: str,
    error: Exception,
    latency_ms: float
) -> None:
    """Logs a failed Groq AI call."""
    logger.error(
        f"[AI Request] ← GROQ ERROR | conv_id: {conversation_id} "
        f"| model: {model} | latency: {latency_ms:.0f}ms "
        f"| {type(error).__name__}: {error}"
    )


def log_prompt_build(
    conversation_id: str,
    prompt_size: int,
    history_turns: int
) -> None:
    """Logs prompt assembly details."""
    logger.info(
        f"[AI Prompt] BUILT | conv_id: {conversation_id} "
        f"| size: {prompt_size} chars | history: {history_turns} turns"
    )


# Aliases — backward compatible
def log_groq_latency(model: str, latency_ms: float, total_tokens: int) -> None:
    throughput = (total_tokens / latency_ms * 1000) if latency_ms > 0 else 0
    logger.info(
        f"[AI Latency] model: {model} | latency: {latency_ms:.2f}ms "
        f"| tokens: {total_tokens} | throughput: {throughput:.1f} tok/s"
    )

def log_ai_call(model_name: str, prompt_tokens: int, completion_tokens: int, latency_ms: float) -> None:
    logger.info(
        f"[AI Telemetry] model: {model_name} | prompt: {prompt_tokens} "
        f"| completion: {completion_tokens} | latency: {latency_ms:.2f}ms"
    )


# ══════════════════════════════════════════════════════════════════
# 3. Database Logging
# ══════════════════════════════════════════════════════════════════

def log_db_connecting(uri_masked: str, db_name: str) -> None:
    logger.info(
        f"[DB] CONNECTING | db: '{db_name}' | uri: {uri_masked}"
    )

def log_db_connected(db_name: str, pool_min: int, pool_max: int, latency_ms: float) -> None:
    logger.info(
        f"[DB] ✓ CONNECTED | db: '{db_name}' "
        f"| pool: {pool_min}–{pool_max} | ping: {latency_ms:.2f}ms"
    )

def log_db_disconnected(db_name: str) -> None:
    logger.info(f"[DB] DISCONNECTED | db: '{db_name}'")

def log_db_error(operation: str, error: Exception) -> None:
    logger.error(
        f"[DB] ERROR | op: '{operation}' | {type(error).__name__}: {error}"
    )

def log_db_fallback(reason: str) -> None:
    logger.warning(
        f"[DB] FALLBACK MODE | reason: {reason}"
    )

def log_db_health(
    db_name: str,
    status: str,
    ping_latency_ms: Optional[float] = None,
    error: Optional[str] = None
) -> None:
    if status == "healthy":
        logger.info(
            f"[DB] HEALTH ✓ | db: '{db_name}' | ping: {ping_latency_ms:.2f}ms"
        )
    else:
        logger.warning(
            f"[DB] HEALTH ✗ | db: '{db_name}' | error: {error}"
        )

def log_crud_insert(collection: str, doc_id: str, latency_ms: float) -> None:
    logger.info(
        f"[DB CRUD] INSERT | collection: '{collection}' "
        f"| id: {doc_id} | latency: {latency_ms:.2f}ms"
    )

def log_crud_find(collection: str, query_key: str, found: bool, latency_ms: float) -> None:
    result = "HIT" if found else "MISS"
    logger.info(
        f"[DB CRUD] FIND {result} | collection: '{collection}' "
        f"| key: {query_key} | latency: {latency_ms:.2f}ms"
    )

def log_crud_update(collection: str, doc_id: str, field: str, latency_ms: float) -> None:
    logger.info(
        f"[DB CRUD] UPDATE | collection: '{collection}' "
        f"| id: {doc_id} | field: {field} | latency: {latency_ms:.2f}ms"
    )

def log_crud_delete(collection: str, doc_id: str, soft: bool, latency_ms: float) -> None:
    mode = "SOFT" if soft else "HARD"
    logger.info(
        f"[DB CRUD] DELETE {mode} | collection: '{collection}' "
        f"| id: {doc_id} | latency: {latency_ms:.2f}ms"
    )

def log_crud_count(collection: str, count: int, latency_ms: float) -> None:
    logger.info(
        f"[DB CRUD] COUNT | collection: '{collection}' "
        f"| count: {count} | latency: {latency_ms:.2f}ms"
    )


# ══════════════════════════════════════════════════════════════════
# 4. RAG Retrieval Logging
# ══════════════════════════════════════════════════════════════════

def log_rag_query(query: str, collection: str, top_k: int) -> None:
    """Logs a RAG vector similarity query."""
    logger.info(
        f"[RAG] QUERY | collection: '{collection}' "
        f"| top_k: {top_k} | query_preview: '{_truncate(query, 60)}'"
    )


def log_rag_results(
    collection: str,
    retrieved: int,
    top_score: float,
    latency_ms: float
) -> None:
    """Logs RAG retrieval results with similarity score and latency."""
    score_flag = " ⚠ LOW RELEVANCE" if top_score < 0.5 else ""
    logger.info(
        f"[RAG] RETRIEVED | collection: '{collection}' "
        f"| docs: {retrieved} | top_score: {top_score:.3f}{score_flag} "
        f"| latency: {latency_ms:.2f}ms"
    )


def log_rag_cache_hit(query_hash: str, collection: str) -> None:
    """Logs a RAG cache hit (retrieved from cache instead of vector DB)."""
    logger.debug(
        f"[RAG] CACHE HIT | collection: '{collection}' | hash: {query_hash}"
    )


def log_rag_miss(query: str, collection: str, reason: str = "no results") -> None:
    """Logs a RAG retrieval miss — no relevant documents found."""
    logger.warning(
        f"[RAG] MISS | collection: '{collection}' | reason: {reason} "
        f"| query: '{_truncate(query, 60)}'"
    )


def log_rag_error(collection: str, error: Exception, latency_ms: float) -> None:
    """Logs a RAG retrieval failure."""
    logger.error(
        f"[RAG] ERROR | collection: '{collection}' "
        f"| latency: {latency_ms:.2f}ms | {type(error).__name__}: {error}"
    )


# ══════════════════════════════════════════════════════════════════
# 5. Tool Call Logging
# ══════════════════════════════════════════════════════════════════

def log_tool_call(
    tool_name: str,
    input_preview: str,
    conversation_id: Optional[str] = None
) -> None:
    """Logs the invocation of any AI tool (sentiment, reasoning, etc.)."""
    conv = f" | conv_id: {conversation_id}" if conversation_id else ""
    logger.info(
        f"[Tool] CALL | tool: '{tool_name}'{conv} "
        f"| input: '{_truncate(input_preview, 80)}'"
    )


def log_tool_result(
    tool_name: str,
    success: bool,
    latency_ms: float,
    result_preview: str = "",
    conversation_id: Optional[str] = None
) -> None:
    """Logs the result of an AI tool call."""
    conv   = f" | conv_id: {conversation_id}" if conversation_id else ""
    status = "✓ OK" if success else "✗ FAILED"
    preview = f" | result: '{_truncate(result_preview, 60)}'" if result_preview else ""
    level  = logging.INFO if success else logging.WARNING
    logger.log(
        level,
        f"[Tool] {status} | tool: '{tool_name}'{conv} "
        f"| latency: {latency_ms:.2f}ms{preview}"
    )


def log_tool_error(
    tool_name: str,
    error: Exception,
    latency_ms: float,
    conversation_id: Optional[str] = None
) -> None:
    """Logs a tool call exception."""
    conv = f" | conv_id: {conversation_id}" if conversation_id else ""
    logger.error(
        f"[Tool] ERROR | tool: '{tool_name}'{conv} "
        f"| latency: {latency_ms:.2f}ms | {type(error).__name__}: {error}"
    )


# ══════════════════════════════════════════════════════════════════
# 6. Error Logging
# ══════════════════════════════════════════════════════════════════

def log_error(
    context: str,
    error: Exception,
    conversation_id: Optional[str] = None
) -> None:
    """Logs an application-level error with full context."""
    conv_part = f" | conv_id: {conversation_id}" if conversation_id else ""
    logger.error(
        f"[Error] {context}{conv_part} | {type(error).__name__}: {error}"
    )


def log_validation_error(field: str, value: Any, reason: str) -> None:
    """Logs a request validation failure."""
    logger.warning(
        f"[Error] VALIDATION | field: '{field}' | value: '{value}' | reason: {reason}"
    )


def log_not_found(resource: str, resource_id: str) -> None:
    """Logs a 404 resource not found event."""
    logger.warning(
        f"[Error] NOT FOUND | resource: '{resource}' | id: '{resource_id}'"
    )


def log_unexpected_error(context: str, error: Exception) -> None:
    """Logs an unexpected/unhandled exception with full stack trace."""
    logger.exception(
        f"[Error] UNEXPECTED | context: {context} | {type(error).__name__}: {error}"
    )


def log_domain_error(
    error_code: str,
    message: str,
    status_code: int,
    path: str
) -> None:
    """Logs a structured domain error (BaseAppException subclass)."""
    level = logging.WARNING if status_code < 500 else logging.ERROR
    logger.log(
        level,
        f"[Error] DOMAIN | code: {error_code} | status: {status_code} "
        f"| path: {path} | msg: {message}"
    )


# ══════════════════════════════════════════════════════════════════
# 7. Response Time Logging
# ══════════════════════════════════════════════════════════════════

def log_request_latency(
    method: str,
    path: str,
    status_code: int,
    latency_ms: float
) -> None:
    """
    Logs HTTP request-level latency.
    Automatically elevates log level if latency exceeds thresholds:
      > 1000ms  → WARNING
      > 5000ms  → ERROR
      4xx/5xx   → WARNING/ERROR
    """
    if latency_ms >= LATENCY_ERROR_MS:
        level = logging.ERROR
        perf  = f"🔴 SLOW"
    elif latency_ms >= LATENCY_WARN_MS:
        level = logging.WARNING
        perf  = f"🟡 SLOW"
    elif status_code >= 500:
        level = logging.ERROR
        perf  = "⚠"
    elif status_code >= 400:
        level = logging.WARNING
        perf  = "⚠"
    else:
        level = logging.INFO
        perf  = "✓"

    logger.log(
        level,
        f"[HTTP] {perf} {method} {path} → {status_code} "
        f"| latency: {latency_ms:.2f}ms"
    )


def log_slow_request(
    method: str,
    path: str,
    latency_ms: float,
    threshold_ms: float
) -> None:
    """Explicitly logs a slow request that exceeded a custom threshold."""
    logger.warning(
        f"[HTTP] SLOW REQUEST | {method} {path} "
        f"| latency: {latency_ms:.0f}ms > threshold: {threshold_ms:.0f}ms"
    )


# ══════════════════════════════════════════════════════════════════
# Decorator — @log_latency
# ══════════════════════════════════════════════════════════════════

def log_latency(operation: str, collection: str = ""):
    """
    Decorator for auto-measuring and logging the execution latency
    of any async repository or service method.

    Usage:
        @log_latency("insert", "conversations")
        async def insert_one(self, ...): ...
    """
    def decorator(func: Callable):
        @wraps(func)
        async def wrapper(*args, **kwargs):
            start = time.perf_counter()
            try:
                result  = await func(*args, **kwargs)
                elapsed = (time.perf_counter() - start) * 1000
                logger.debug(
                    f"[DB CRUD] {operation.upper()} "
                    f"{'| ' + collection + ' ' if collection else ''}"
                    f"| {elapsed:.2f}ms"
                )
                return result
            except Exception as e:
                elapsed = (time.perf_counter() - start) * 1000
                logger.error(
                    f"[DB CRUD] ERROR {operation.upper()} "
                    f"{'| ' + collection + ' ' if collection else ''}"
                    f"| {elapsed:.2f}ms | {type(e).__name__}: {e}"
                )
                raise
        return wrapper
    return decorator


# ══════════════════════════════════════════════════════════════════
# Backward-Compatible Aliases
# ══════════════════════════════════════════════════════════════════

def log_conversation(
    conversation_id: str,
    message_id: str,
    prompt_size: int,
    execution_time_ms: float,
    prompt_tokens: int,
    completion_tokens: int,
    total_tokens: int,
    error: str = None
) -> None:
    """Logs conversation telemetry — backward-compatible alias."""
    err_suffix = f" | ERROR: {error}" if error else ""
    logger.info(
        f"[Conversation] conv_id: {conversation_id} | msg_id: {message_id} "
        f"| prompt: {prompt_size} chars | latency: {execution_time_ms:.2f}ms "
        f"| tokens: prompt={prompt_tokens} completion={completion_tokens} total={total_tokens}"
        f"{err_suffix}"
    )


# ══════════════════════════════════════════════════════════════════
# Private Helpers
# ══════════════════════════════════════════════════════════════════

def _mask_email(email: str) -> str:
    """Masks the local part of an email for safe logging: user@x.com → u***@x.com"""
    if not email or "@" not in email:
        return "***"
    local, domain = email.split("@", 1)
    return f"{local[0]}***@{domain}" if len(local) > 1 else f"***@{domain}"


def _truncate(text: str, max_len: int) -> str:
    """Truncates text for safe log line length."""
    if not text:
        return ""
    text = str(text).replace("\n", " ").strip()
    return text[:max_len] + "…" if len(text) > max_len else text

File: server/config/logger.js
/**
 * Enterprise CX Guardian AI — Node.js Centralized Logger
 * =======================================================
 *
 * 7 Logging Categories (matches FastAPI service):
 *   1. auth       — login, logout, token events, failures
 *   2. ai         — AI proxy requests, model, tokens, latency
 *   3. db         — MongoDB queries, connections, CRUD timing
 *   4. rag        — RAG retrieval calls (future)
 *   5. tool       — Sentiment, reasoning, recommendation tool calls
 *   6. error      — Domain errors, validation, unexpected exceptions
 *   7. http       — Response time, per-route latency, slow request alerts
 *
 * Transports:
 *   - Console (colored, dev-friendly)
 *   - logs/error.log (error-level only, JSON)
 *   - logs/combined.log (all levels, JSON, structured)
 *   - logs/audit.log (auth events only, JSON — for compliance)
 */

import winston from 'winston';
import path from 'path';
import { fileURLToPath } from 'url';
import * as env from './env.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname  = path.dirname(__filename);
const LOG_DIR    = path.join(__dirname, '../logs');

// ── Latency thresholds ─────────────────────────────────────────────
const LATENCY_WARN_MS  = 1000;   // 1 second  → WARN
const LATENCY_ERROR_MS = 5000;   // 5 seconds → ERROR

// ══════════════════════════════════════════════════════════════════
// Winston Setup
// ══════════════════════════════════════════════════════════════════

const levels = { error: 0, warn: 1, info: 2, http: 3, debug: 4 };
const colors = { error: 'red', warn: 'yellow', info: 'green', http: 'magenta', debug: 'white' };
winston.addColors(colors);

const getLevel = () => env.NODE_ENV === 'development' ? 'debug' : 'info';

const fileFormat = winston.format.combine(
  winston.format.timestamp({ format: 'YYYY-MM-DD HH:mm:ss' }),
  winston.format.errors({ stack: true }),
  winston.format.splat(),
  winston.format.json()
);

const consoleFormat = winston.format.combine(
  winston.format.colorize({ all: true }),
  winston.format.timestamp({ format: 'YYYY-MM-DD HH:mm:ss' }),
  winston.format.printf(({ timestamp, level, message }) =>
    `[${timestamp}] [${level}]: ${message}`
  )
);

const _base = winston.createLogger({
  level: getLevel(),
  levels,
  transports: [
    new winston.transports.Console({ format: consoleFormat }),
    new winston.transports.File({ filename: `${LOG_DIR}/error.log`,    level: 'error', format: fileFormat }),
    new winston.transports.File({ filename: `${LOG_DIR}/combined.log`, level: 'debug', format: fileFormat }),
  ]
});

// Separate audit transport for auth events (compliance trail)
const _audit = winston.createLogger({
  level: 'info',
  levels,
  transports: [
    new winston.transports.File({ filename: `${LOG_DIR}/audit.log`, format: fileFormat }),
    new winston.transports.Console({ format: consoleFormat })
  ]
});

// ══════════════════════════════════════════════════════════════════
// Private Helpers
// ══════════════════════════════════════════════════════════════════

const _maskEmail = (email = '') => {
  if (!email || !email.includes('@')) return '***';
  const [local, domain] = email.split('@');
  return `${local[0]}***@${domain}`;
};

const _truncate = (text = '', max = 80) => {
  const s = String(text).replace(/\n/g, ' ').trim();
  return s.length > max ? s.slice(0, max) + '…' : s;
};

const _ts = () => new Date().toISOString();

// ══════════════════════════════════════════════════════════════════
// 1. Authentication Logging
// ══════════════════════════════════════════════════════════════════

const auth = {
  attempt:      (email, method = 'password') =>
    _audit.info(`[Auth] LOGIN ATTEMPT | email: ${_maskEmail(email)} | method: ${method}`),

  success:      (userId, email, tokenType = 'access') =>
    _audit.info(`[Auth] ✓ LOGIN SUCCESS | user_id: ${userId} | email: ${_maskEmail(email)} | token: ${tokenType}`),

  failure:      (email, reason, ip = 'unknown') =>
    _audit.warn(`[Auth] ✗ LOGIN FAILED | email: ${_maskEmail(email)} | reason: ${reason} | ip: ${ip}`),

  logout:       (userId) =>
    _audit.info(`[Auth] LOGOUT | user_id: ${userId}`),

  register:     (email, userId) =>
    _audit.info(`[Auth] REGISTER | user_id: ${userId} | email: ${_maskEmail(email)}`),

  tokenRefresh: (userId, success, reason = '') =>
    _audit.info(`[Auth] TOKEN REFRESH | user_id: ${userId} | ${success ? '✓ SUCCESS' : `✗ FAILED (${reason})`}`),

  tokenInvalid: (reason, ip = 'unknown') =>
    _audit.warn(`[Auth] INVALID TOKEN | reason: ${reason} | ip: ${ip}`),
};

// ══════════════════════════════════════════════════════════════════
// 2. AI Request Logging
// ══════════════════════════════════════════════════════════════════

const ai = {
  requestStart: (convId, model, historyTurns, promptChars) =>
    _base.info(`[AI Request] → PROXY | conv_id: ${convId} | model: ${model} | history: ${historyTurns} turns | prompt: ${promptChars} chars`),

  requestSuccess: (convId, model, tokens = {}, latencyMs) => {
    const { prompt = 0, completion = 0, total = 0 } = tokens;
    const throughput = latencyMs > 0 ? ((total / latencyMs) * 1000).toFixed(0) : 0;
    const level = latencyMs >= LATENCY_ERROR_MS ? 'error' : latencyMs >= LATENCY_WARN_MS ? 'warn' : 'info';
    _base[level](`[AI Request] ← OK | conv_id: ${convId} | model: ${model} | tokens: prompt=${prompt} completion=${completion} total=${total} | latency: ${latencyMs.toFixed(0)}ms | throughput: ${throughput} tok/s`);
  },

  requestError: (convId, model, error, latencyMs) =>
    _base.error(`[AI Request] ← ERROR | conv_id: ${convId} | model: ${model} | latency: ${latencyMs?.toFixed(0)}ms | ${error?.message || error}`),

  proxyCall: (endpoint, method, latencyMs, status) =>
    _base.info(`[AI Proxy] ${method} ${endpoint} → ${status} | latency: ${latencyMs?.toFixed(0)}ms`),
};

// ══════════════════════════════════════════════════════════════════
// 3. Database Logging
// ══════════════════════════════════════════════════════════════════

const db = {
  connecting:   (uri, dbName) =>
    _base.info(`[DB] CONNECTING | db: '${dbName}' | uri: ${uri}`),

  connected:    (dbName, latencyMs) =>
    _base.info(`[DB] ✓ CONNECTED | db: '${dbName}' | ping: ${latencyMs?.toFixed(2)}ms`),

  disconnected: (dbName) =>
    _base.info(`[DB] DISCONNECTED | db: '${dbName}'`),

  error:        (operation, err) =>
    _base.error(`[DB] ERROR | op: '${operation}' | ${err?.message || err}`),

  health:       (dbName, healthy, latencyMs, err) =>
    healthy
      ? _base.info(`[DB] HEALTH ✓ | db: '${dbName}' | ping: ${latencyMs?.toFixed(2)}ms`)
      : _base.warn(`[DB] HEALTH ✗ | db: '${dbName}' | error: ${err}`),

  query:        (collection, operation, filter, latencyMs) =>
    _base.debug(`[DB CRUD] ${operation.toUpperCase()} | collection: '${collection}' | filter: ${JSON.stringify(filter)} | latency: ${latencyMs?.toFixed(2)}ms`),

  insert:       (collection, id, latencyMs) =>
    _base.info(`[DB CRUD] INSERT | collection: '${collection}' | id: ${id} | latency: ${latencyMs?.toFixed(2)}ms`),

  update:       (collection, id, field, latencyMs) =>
    _base.info(`[DB CRUD] UPDATE | collection: '${collection}' | id: ${id} | field: ${field} | latency: ${latencyMs?.toFixed(2)}ms`),

  delete:       (collection, id, soft, latencyMs) =>
    _base.info(`[DB CRUD] DELETE ${soft ? 'SOFT' : 'HARD'} | collection: '${collection}' | id: ${id} | latency: ${latencyMs?.toFixed(2)}ms`),

  find:         (collection, key, found, latencyMs) =>
    _base.info(`[DB CRUD] FIND ${found ? 'HIT' : 'MISS'} | collection: '${collection}' | key: ${key} | latency: ${latencyMs?.toFixed(2)}ms`),
};

// ══════════════════════════════════════════════════════════════════
// 4. RAG Retrieval Logging
// ══════════════════════════════════════════════════════════════════

const rag = {
  query:   (query, collection, topK) =>
    _base.info(`[RAG] QUERY | collection: '${collection}' | top_k: ${topK} | query: '${_truncate(query, 60)}'`),

  results: (collection, count, topScore, latencyMs) =>
    _base.info(`[RAG] RETRIEVED | collection: '${collection}' | docs: ${count} | top_score: ${topScore?.toFixed(3)} | latency: ${latencyMs?.toFixed(2)}ms`),

  miss:    (query, collection, reason = 'no results') =>
    _base.warn(`[RAG] MISS | collection: '${collection}' | reason: ${reason} | query: '${_truncate(query, 60)}'`),

  error:   (collection, err, latencyMs) =>
    _base.error(`[RAG] ERROR | collection: '${collection}' | latency: ${latencyMs?.toFixed(2)}ms | ${err?.message || err}`),
};

// ══════════════════════════════════════════════════════════════════
// 5. Tool Call Logging
// ══════════════════════════════════════════════════════════════════

const tool = {
  call:   (toolName, inputPreview, convId) =>
    _base.info(`[Tool] CALL | tool: '${toolName}'${convId ? ` | conv_id: ${convId}` : ''} | input: '${_truncate(inputPreview, 80)}'`),

  result: (toolName, success, latencyMs, resultPreview = '', convId) => {
    const status  = success ? '✓ OK' : '✗ FAILED';
    const preview = resultPreview ? ` | result: '${_truncate(resultPreview, 60)}'` : '';
    const conv    = convId ? ` | conv_id: ${convId}` : '';
    _base[success ? 'info' : 'warn'](`[Tool] ${status} | tool: '${toolName}'${conv} | latency: ${latencyMs?.toFixed(2)}ms${preview}`);
  },

  error:  (toolName, err, latencyMs, convId) =>
    _base.error(`[Tool] ERROR | tool: '${toolName}'${convId ? ` | conv_id: ${convId}` : ''} | latency: ${latencyMs?.toFixed(2)}ms | ${err?.message || err}`),
};

// ══════════════════════════════════════════════════════════════════
// 6. Error Logging
// ══════════════════════════════════════════════════════════════════

const error = {
  domain:      (errorCode, message, statusCode, path) => {
    const level = statusCode >= 500 ? 'error' : 'warn';
    _base[level](`[Error] DOMAIN | code: ${errorCode} | status: ${statusCode} | path: ${path} | msg: ${message}`);
  },

  validation:  (field, value, reason) =>
    _base.warn(`[Error] VALIDATION | field: '${field}' | value: '${value}' | reason: ${reason}`),

  notFound:    (resource, id) =>
    _base.warn(`[Error] NOT FOUND | resource: '${resource}' | id: '${id}'`),

  unexpected:  (context, err) =>
    _base.error(`[Error] UNEXPECTED | context: ${context} | ${err?.stack || err}`),

  middleware:  (path, method, err) =>
    _base.error(`[Error] MIDDLEWARE | ${method} ${path} | ${err?.message || err}`),
};

// ══════════════════════════════════════════════════════════════════
// 7. HTTP / Response Time Logging
// ══════════════════════════════════════════════════════════════════

const http = {
  request: (method, path, statusCode, latencyMs) => {
    let level, perf;
    if      (latencyMs >= LATENCY_ERROR_MS) { level = 'error'; perf = '🔴 SLOW'; }
    else if (latencyMs >= LATENCY_WARN_MS)  { level = 'warn';  perf = '🟡 SLOW'; }
    else if (statusCode >= 500)             { level = 'error'; perf = '⚠';      }
    else if (statusCode >= 400)             { level = 'warn';  perf = '⚠';      }
    else                                    { level = 'info';  perf = '✓';      }
    _base[level](`[HTTP] ${perf} ${method} ${path} → ${statusCode} | latency: ${latencyMs?.toFixed(2)}ms`);
  },

  slow: (method, path, latencyMs, thresholdMs) =>
    _base.warn(`[HTTP] SLOW REQUEST | ${method} ${path} | latency: ${latencyMs?.toFixed(0)}ms > threshold: ${thresholdMs}ms`),
};

// ══════════════════════════════════════════════════════════════════
// Exports
// ══════════════════════════════════════════════════════════════════

const logger = {
  // Category namespaces
  auth,
  ai,
  db,
  rag,
  tool,
  error,
  http,

  // Direct winston pass-through for generic logging
  info:  (...args) => _base.info(...args),
  warn:  (...args) => _base.warn(...args),
  debug: (...args) => _base.debug(...args),
  error: (...args) => _base.error(...args),
  http:  (...args) => _base.http(...args),
};

export default logger;
